### PR TITLE
Create new account during transferring if the to_address is not registered.

### PR DESCRIPTION
**Pull request by magicalne**
In this pr, a new EoA account is created during native token transferring if the `to_address` is not registered.

And fix the test due to updates of recover account from godwoken.

## Related PR
- https://github.com/nervosnetwork/godwoken-polyjuice/pull/173

### CODE REVIEW

File: c/polyjuice.h [hunk 1]
```diff
@@ -1066,7 +1082,30 @@ int handle_native_token_transfer(gw_context_t* ctx, uint32_t from_id,
...
+    uint8_t script_args[eoa_script_args_len];
+    memcpy(script_args, g_rollup_script_hash, 32);
+    memcpy(script_args + 32, g_eoa_transfer_to_address.bytes, 20);
+    mol_seg_t new_script_seg;
+    ret = build_script(eoa_type_hash, g_script_hash_type, script_args,
+                       eoa_script_args_len, &new_script_seg);
+    if (ret != 0) {
+      return ret;
+    }
+    uint8_t script_hash[32];
```
Comment (Flouse) [inline c/polyjuice.h:1103]: What is `script_hash` declared for?

Comment (magicalne) [reply c/polyjuice.h:1103]: A mistake. Should be removed.

File: c/polyjuice.h [hunk 2]
```diff
@@ -1020,6 +1020,22 @@ int handle_transfer(gw_context_t* ctx,
...
 
+int load_eoa_type_hash(gw_context_t* ctx, uint8_t eoa_type_hash[GW_KEY_BYTES]) {
+  mol_seg_t rollup_config_seg;
+  rollup_config_seg.ptr = ctx->rollup_config;
+  rollup_config_seg.size = ctx->rollup_config_size;
+
+  mol_seg_t allowed_eoa_list_seg =
+      MolReader_RollupConfig_get_allowed_eoa_type_hashes(&rollup_config_seg);
+  mol_seg_res_t allowed_type_hash_res =
+      MolReader_AllowedTypeHashVec_get(&allowed_eoa_list_seg, 0);
```
Comment (Flouse) [inline c/polyjuice.h:1031]: allowed eoa: 0: unknown, 1: eth, 2: tron

I think we need to check [AllowedTypeHash.type_](https://github.com/nervosnetwork/godwoken/blob/develop/crates/types/schemas/godwoken.mol#L40)

## How to load eth_eoa_type_hash?
refer to https://github.com/nervosnetwork/godwoken-scripts/blob/master/c/gw_eth_addr_reg.h#L149-L162

File: c/polyjuice.h [hunk 3]
```diff
@@ -1484,14 +1521,25 @@ int run_polyjuice() {
...
       value_ptr[i] = msg.value.bytes[31 - i];
     }
+
+    uint64_t gas_used = min_gas;
+    // charge gas fee of account creation if to_address is not registered
+    gw_reg_addr_t to_addr = new_reg_addr(g_eoa_transfer_to_address.bytes);
+    uint8_t to_script_hash[GW_KEY_BYTES] = {0};
+    ret = context.sys_get_script_hash_by_registry_address(&context, &to_addr, to_script_hash);
+    if (ret == GW_ERROR_NOT_FOUND) {
+        gas_used += NEW_ACCOUNT_GAS;
```
Comment (keroro520) [inline c/polyjuice.h:1531]: The `*gasUsed` should be passed into the `handle_native_token_transfer` function and increased immediately when a new account is created.

File: c/polyjuice.h (hunk 2 above)
Comment (keroro520) [reply c/polyjuice.h:1031]: > [allowed eoa](https://github.com/nervosnetwork/godwoken-scripts/blob/master/c/gw_def.h#L53): 0: unknown, 1: eth, 2: tron
> 
> ## How to load eth_eoa_type_hash?
> I think we need to interate allowed_eoa_type_hashes and find the correct [AllowedTypeHash.type_](https://github.com/nervosnetwork/godwoken/blob/develop/crates/types/schemas/godwoken.mol#L40).
> 
> refer to https://github.com/nervosnetwork/godwoken-scripts/blob/master/c/gw_eth_addr_reg.h#L149-L162

+1

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: c/polyjuice.h (hunk 2 above)
Comment (magicalne) [reply c/polyjuice.h]: Done!

File: c/polyjuice.h (hunk 3 above)
Comment (magicalne) [reply c/polyjuice.h]: Done!

**Review by zeroqn — DISMISSED** [review]

**Review by keroro520 — DISMISSED** [review]

File: c/polyjuice.h [hunk 4]
```diff
@@ -1020,8 +1020,41 @@ int handle_transfer(gw_context_t* ctx,
...
+  rollup_config_seg.size = ctx->rollup_config_size;
+
+  mol_seg_t allowed_eoa_list_seg =
+      MolReader_RollupConfig_get_allowed_eoa_type_hashes(&rollup_config_seg);
+  uint32_t len = MolReader_AllowedTypeHashVec_length(&allowed_eoa_list_seg);
+  for (uint32_t i = 0; i < len; i++) {
+    mol_seg_res_t allowed_type_hash_res =
+        MolReader_AllowedTypeHashVec_get(&allowed_eoa_list_seg, i);
+
+    if (allowed_type_hash_res.errno != MOL_OK) {
```
Comment (Flouse) [inline c/polyjuice.h:1037]: Maybe we function [is_errno_ok](https://github.com/nervosnetwork/godwoken-polyjuice/blob/1.3.0/c/polyjuice_utils.h#L60-L66) is a better choice here.

File: c/polyjuice.h [hunk 5]
```diff
@@ -1020,8 +1020,41 @@ int handle_transfer(gw_context_t* ctx,
...
+        MolReader_AllowedTypeHash_get_type_(&allowed_type_hash_res.seg);
+    if (*(uint8_t *)type_seg.ptr == GW_ALLOWED_EOA_ETH) {
+      mol_seg_t eth_lock_code_hash_seg =
+          MolReader_AllowedTypeHash_get_hash(&allowed_type_hash_res.seg);
+      memcpy(eoa_type_hash, eth_lock_code_hash_seg.ptr, 32);
+      return 0;
+    }
+  }
+  ckb_debug("Cannot find EoA type hash of ETH.");
+  return -1;
```
Comment (Flouse) [inline c/polyjuice.h:1051]:
```suggestion
  return FATAL_POLYJUICE;
```

**Review by Flouse — APPROVED** [review]

**Review by jjyr — CHANGES_REQUESTED** [review]

File: c/polyjuice.h (hunk 4 above)
Comment (magicalne) [reply c/polyjuice.h]: Done

File: c/polyjuice.h [hunk 6]
```diff
@@ -1066,7 +1099,34 @@ int handle_native_token_transfer(gw_context_t* ctx, uint32_t from_id,
...
+    if (ret != 0) {
+      return ret;
+    }
+    uint32_t new_account_id;
+    ret = ctx->sys_create(ctx, new_script_seg.ptr, new_script_seg.size,
+                          &new_account_id);
+    if (ret != 0) {
+      ckb_debug("[handle_native_token_transfer] create new account failed.");
+      return ret;
+    }
```
[the remark this answers is not in the document]
Comment (magicalne) [reply c/polyjuice.h:1123]: Done!

**Review by Flouse — APPROVED** [review]

**Review by zeroqn — APPROVED** [review]

File: c/polyjuice.h [hunk 7]
```diff
@@ -1066,7 +1095,49 @@ int handle_native_token_transfer(gw_context_t* ctx, uint32_t from_id,
...
+    uint8_t account_script_hash[32] = {0};
+    ret = ctx->sys_get_script_hash_by_account_id(ctx, new_account_id,
+                                                 account_script_hash);
+    if (ret != 0) {
+      ckb_debug("[handle_native_token_transfer] failed to get created eth account script hash");
+      return ret;
+    }
+
+    ret = gw_register_eth_address(ctx, account_script_hash);
+    ckb_debug("hit here");
```
Comment (jjyr) [inline c/polyjuice.h:1133]: Remove debug output?

File: c/polyjuice.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jjyr) [reply c/polyjuice.h:1123]: Please remover debug output

**Review by Flouse — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

**Pull request merged**